samcli/lib/cookiecutter/template.py

```diff
@@ -0,0 +1,164 @@
+        templates, it is better to be extracted to a plugin that can then be plugged in to each of these templates.
+
+    Methods
+    -------
+    run_interactive_flows() -> Dict:
+        runs the interactive_flows(questions) one be one then combines and returns the answers in one dictionary
+        in the form of {question.key: answer}
+    generate_project(context) -> None:
+        runs the preprocessors chain passing in the given context to the first processor which process it then produce
+        a new context which is passed to the next preprocessor and so on.
+        After that, it execute the cookiecutter generator to generate the project with the final preprocessed context.
+        Finally, it runs the postprocessors chain with the same manner like the preprocessors.
+    """
+
+    def __init__(
+        self,
+        location: str,
+        interactive_flows: Optional[List[InteractiveFlow]] = None,
+        preprocessors: Optional[List[Processor]] = None,
+        postprocessors: Optional[List[Processor]] = None,
+        plugins: Optional[List[Plugin]] = None,
+    ):
+        """
+        Initialize the class
+
+        Parameters
+        ----------
+        location: str
+            Template location (git, mercurial, http(s), path)
+        interactive_flows: Optional[List[InteractiveFlow]]
+            An optional series of interactive questions to be asked to the user where the answers are used to fulfill
+            the values of the cookiecutter.json keys, i.e. the cookiecutter context
+        preprocessors: Optional[List[Processor]]
+            An optional series of processes to be executed before generating the cookiecutter project. These processes
+            are usually used to manipulate the cookiecutter context or doing some bootstrapping before generating
+            the cookiecutter project.
+        postprocessors: Optional[List[Processor]]
+            An optional series of processes to be executed after generating the cookiecutter project. These processes
+            are usually used to clean up resources and to print messages to the user after successfully generating
+            the cookiecutter project.
+        plugins: Optional[List[Plugin]]
+            An optional series of plugins to be plugged in. A plugin defines its own interactive_flow, preprocessor
+            and postprocessor. A plugin is a sub-set of the template, if there is a common behavior among multiple
+            templates, it is better to be extracted to a plugin that can then be plugged in to each of these templates.
+        """
+        self._location = location
+        self._interactive_flows = interactive_flows or []
+        self._preprocessors = preprocessors or []
+        self._postprocessors = postprocessors or []
+        self._plugins = plugins or []
+        for plugin in self._plugins:
+            if plugin.interactive_flow:
+                self._interactive_flows.append(plugin.interactive_flow)
```

Comment:
does this control order of execution? should plugin interactive flow be the first thing? similairly for pre and post processors? or does it not matter?

Reply:
A standard flow is as following:
1. InteractiveFlow: interactive questionnaire that gather the user's answers and create the cookiecutter context 
2. Preprocessing: a template-defined logic to process the context, for example the preprocessing of the pipeline template(coming next) is to create the required AWS resources if the user didn't provide it
3. Cookiecutter: generating the cookiecutter project form the template's files
4. postprocessing: a template-defined logic. for example the pipeline templates will use the post-processing to print instructions on what resources it created on behalf of the users.

So a template defines an interactive-flow, pre-processors and post-processors.

Now, multiple templates may have a similar flow and processing logic, and this is where the **plugin** concept come into place. We can extract the common interactive-flow, preprocessor and postprocessor to a plugin, then plug it into each of these template which will just append these three components to the corresponding template's components.

Regarding the order; it will be the template's own component then the plugins' components in the same order they are plugged in.

Example:
t = Template(interactive_flows=[f1, f2], preprocessors=[p1, p2,], postprocessors=[p3, p4], plugins=[g1,g2])
the order of execution of each component will be as following:
1. interactive_flows: f1, f2, g1.interactive_flow, g2.interactive_flow
2. preprocesspors: p1, p2, g1.preprocesspor, g2.preprocesspor
3. postprocesspors: p3, p4, g1.postprocesspor, g2.postprocesspor

Reply:
Thanks for the detailed explanation!